fresnel/cpu/GeometryPolygon.cc

```diff
@@ -0,0 +1,243 @@
+
+    // set default material
+    setMaterial(Material(RGB<float>(1,0,1)));
+    setOutlineMaterial(Material(RGB<float>(0,0,0), 1.0f));
+
+    // allocate buffer data
+    m_position = std::shared_ptr< Array< vec2<float> > >(new Array< vec2<float> >(N));
+    m_angle = std::shared_ptr< Array< float > >(new Array< float >(N));
+    m_color = std::shared_ptr< Array< RGB<float> > >(new Array< RGB<float> >(N));
+
+    // copy the vertices from the numpy array to internal storage
+    pybind11::buffer_info info = vertices.request();
+
+    if (info.ndim != 2)
+        throw std::runtime_error("vertices must be a 2-dimensional array");
+
+    if (info.shape[1] != 2)
+        throw std::runtime_error("vertices must be a Nvert by 2 array");
+
+    float *verts_f = (float *)info.ptr;
+
+    for (unsigned int i = 0; i < info.shape[0]; i++)
+        {
+        vec2<float> p0(verts_f[i*2], verts_f[i*2+1]);
+
+        m_vertices.push_back(p0);
+
+        // precompute radius in the xy plane
+        m_radius = std::max(m_radius, sqrtf(dot(p0,p0)));
+        }
+    // pad the radius with the rounding radius
+    m_radius += m_rounding_radius;
+
+    // register functions for embree
+    rtcSetGeometryUserData(m_geometry, this);
+    m_device->checkError();
+    rtcSetGeometryBoundsFunction(m_geometry, &GeometryPolygon::bounds, NULL);
+    m_device->checkError();
+    rtcSetGeometryIntersectFunction(m_geometry, &GeometryPolygon::intersect);
+    m_device->checkError();
+
+    rtcCommitGeometry(m_geometry);
+    m_device->checkError();
+
+    m_valid = true;
+    }
+
+GeometryPolygon::~GeometryPolygon()
+    {
+    }
+
+/*! Compute the bounding box of a given primitive
+
+    \param ptr Pointer to a GeometryPolygon instance
+    \param item Index of the primitive to compute the bounding box of
+    \param bounds_o Output bounding box
+*/
+void GeometryPolygon::bounds(const struct RTCBoundsFunctionArguments *args)
+    {
+    GeometryPolygon *geom = (GeometryPolygon*)args->geometryUserPtr;
+    vec2<float> p2 = geom->m_position->get(args->primID);
+    vec3<float> p(p2.x, p2.y, 0.0f);
+
+    RTCBounds& bounds_o = *args->bounds_o;
+    bounds_o.lower_x = p.x - geom->m_radius;
+    bounds_o.lower_y = p.y - geom->m_radius;
+    bounds_o.lower_z = p.z - 1e-5;
+
+    bounds_o.upper_x = p.x + geom->m_radius;
+    bounds_o.upper_y = p.y + geom->m_radius;
+    bounds_o.upper_z = p.z + 1e-5;
+    }
+
+//! Test if a point is inside a polygon
+/*! \param min_d  [out] minimum distance from p to the polygon edge
+    \param p Point
+    \param verts Polygon vertices
+
+    \returns true if the point is inside the polygon
+
+    \note \a p is *in the polygon's reference frame!*
+
+    \ingroup overlap
+*/
+inline bool is_inside(float &min_d, const vec2<float>& p, const std::vector< vec2<float> >& verts)
+    {
+    // code for concave test from: http://alienryderflex.com/polygon/
+    unsigned int nvert = verts.size();
+    min_d = FLT_MAX;
+
+    unsigned int  i, j=nvert-1;
```

Comment:
Formatting is kind of weird throughout this function - there are extra spaces on this line and several lines here and below with missing spaces around equals signs used for assignment. Would be easier to read if those were fixed.